### PR TITLE
feat(properties-timeline): Add backend of properties timeline

### PR DESCRIPTION
**Pull request by Twixes**
## Problem

Persons-/groups-on-events makes person/group properties point-in-time, which is not clear to users. In particular, it's not clear from an insight how those properties change over time, which makes it opaque _why_ an actor might be counted at one point in time, but not at another.

## Changes

This introduces what is basically a new insight (lite): Properties Timeline.
A properties timeline is a list of points at which the person's properties changed **in a relevant way**. In the context of an insight, "relevant way" means that one of the property actually used for filtering in that insight (or even that particular series, for Trends) must have changed. Changes to other properties are ignored, since that's just noise in this context.

In a follow-up PR: support for groups.

## How did you test this code?

Added a new test suite, `TestPropertiesTimeline`.

### CODE REVIEW

**Comment by neilkakkar** [comment]
Random feedback (if I'm understanding the query correctly):

You can do better perf-wise here by not using window functions, since you're dealing with events for a single person. Instead, use `GROUP BY` and use array functions to manipulate things.

For example: https://clickhouse.com/docs/en/sql-reference/functions/array-functions/#arrayenumeratedensearr gives you an array like: [1,1,1,1,2,2,3,3,3,1,1,1,4,4] (each number is a property state, it increments when it changes; it goes back to 1 when new changed property is same as the old one),

and then you have to sum all consecutive numbers that are the same to get the event counts per unique property.

---

Given it's events for only one person though, I don't expect perf to be a big concern 🤷

File: posthog/api/test/__snapshots__/test_person.ambr [hunk 1]
```diff
@@ -116,3 +116,335 @@
...
+               person_properties AS properties,
+               tuple(replaceRegexpAll(JSONExtractRaw(person_properties, 'bar'), '^"|"$', '')) AS relevant_property_values,
+               lagInFrame(tuple(replaceRegexpAll(JSONExtractRaw(person_properties, 'bar'), '^"|"$', ''))) OVER person_events AS previous_relevant_property_values,
+                                                                                                               row_number() OVER person_events AS start_event_number
+        FROM (
+                (SELECT e.timestamp as timestamp,
+                        e."person_properties" as "person_properties"
+                 FROM events e PREWHERE team_id = 2
+                 AND e.person_id = '00000000-0000-4000-8000-000000000001'
+                 AND (event = '$pageview')
```
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/test/__snapshots__/test_person.ambr]: Just made this a `WHERE` since soon™ we'll have `optimize_move_to_prewhere` enabled anyway

[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/test/__snapshots__/test_person.ambr]: It's not clear to me why property filters are a _problem_ for prewhere though, can you explain? 👀

**Comment by Twixes** [comment]
I initially used a `GROUP BY` approach, but that didn't work when a property value was changed _back_ to one seen before – the resulting timeline was incomplete this way. As for array functions, I don't quite see how this'd work with `arrayEnumerateDense`, because we need to operate on rows to preserve the properties and timestamp data. I may totally be missing something though, let me know if you meant something different @neilkakkar.

File: posthog/api/test/__snapshots__/test_person.ambr (hunk 1 above)
[the remark this answers is not in the document]
Comment (Twixes) [reply posthog/api/test/__snapshots__/test_person.ambr]: :+1: Makes sense, will rely on the optimizer then

File: posthog/models/entity/util.py [hunk 2]
```diff
@@ -7,29 +7,44 @@
 
 
 def get_entity_filtering_params(
```
Comment (macobo) [inline posthog/models/entity/util.py:9]: Is the behavioral change here an unrelated fix?

File: posthog/models/filters/properties_timeline_filter.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
+from posthog.models.filters.base_filter import BaseFilter
+from posthog.models.filters.mixins.common import DateMixin, EntitiesMixin
+from posthog.models.filters.mixins.groups import GroupsAggregationMixin
+from posthog.models.filters.mixins.property import PropertyMixin
+
+
+class PropertiesTimelineFilter(
```
Comment (macobo) [inline posthog/models/filters/properties_timeline_filter.py:7]: Worth chatting with @mariusandra and making this use data exploration only instead?

File: posthog/queries/properties_timeline/properties_timeline.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+from .properties_timeline_event_query import PropertiesTimelineEventQuery
+
+
+class PropertiesTimelinePoint(TypedDict):
+    timestamp: str
+    properties: Dict[str, Any]
+    relevant_events_since_previous_point: int
+
+
+PROPERTIES_TIMELINE_SQL = """
```
Comment (macobo) [inline posthog/queries/properties_timeline/properties_timeline.py:19]: I don't understand this query (probably because I'm lacking context on what problem `relevant_events_since_previous_point` solves.

- Have you tried this query in production for large teams?
- Are there other ways to solve the same problem (e.g. by doing multiple queries)

File: posthog/api/test/test_person_properties_timeline.py [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+            timeline,
+            [
+                {
+                    "properties": {"foo": "abc", "bar": 456},  # Initial bar
+                    "relevant_events_since_previous_point": 1,  # 1 in first entry means this person existed pre-range
+                    "timestamp": "2019-12-27T00:00:00Z",
+                },
+                {
+                    "properties": {"foo": "abc", "bar": 123},  # Changed bar
+                    "relevant_events_since_previous_point": 1,
```
Comment (macobo) [inline posthog/api/test/test_person_properties_timeline.py:133]: what will this property be used by in the frontend?

File: posthog/queries/event_query/event_query.py [hunk 6]
```diff
@@ -141,7 +151,9 @@ def _does_cohort_need_persons(self, prop: Property) -> bool:
         return False
 
     @cached_property
-    def _person_query(self):
+    def _person_query(self) -> PersonQuery:
+        if isinstance(self._filter, PropertiesTimelineFilter):
```
Comment (macobo) [inline posthog/queries/event_query/event_query.py:155]: This feels clunky in that we are inheriting but not really. This mega-class just got more exceptional than anywhere else.

Could we create a PropertiesTimelineEventQuery and override these functions to raise there instead of hoisting its concerns elsewhere?

File: posthog/queries/query_date_range.py [hunk 7]
```diff
@@ -221,4 +247,6 @@ def should_round(self):
         return round_interval
 
     def is_hourly(self, target):
-        return self._filter.interval == "hour" or (target and isinstance(target, str) and "h" in target)
+        return getattr(self._filter, "interval", None) == "hour" or (
```
Comment (macobo) [inline posthog/queries/query_date_range.py:250]: Needing to use getattr with filters makes the code much harder to work with. Can we avoid this somehow?

File: posthog/queries/query_date_range.py [hunk 8]
```diff
@@ -14,9 +14,28 @@
 from posthog.utils import DEFAULT_DATE_FROM_DAYS
 
 
-# Assume that any date being sent from the client is timezone aware according to the timezone that the team has set
 class QueryDateRange:
```
Comment (macobo) [inline posthog/queries/query_date_range.py:17]: Rather than extend this already complicated base class to do even more - could we use inheritance/centralize the time-line specific functionality to timeline?

File: posthog/queries/properties_timeline/properties_timeline.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        lagInFrame(start_event_number) OVER person_points AS previous_start_event_number
+    FROM (
+        SELECT
+            timestamp,
+            person_properties AS properties,
+            is_pre_range,
+            {crucial_property_columns} AS relevant_property_values, -- TODO make this dynamic
+            lagInFrame({crucial_property_columns}) OVER person_events AS previous_relevant_property_values,
+            row_number() OVER person_events AS start_event_number
+        FROM ({event_query})
```
Comment (macobo) [inline posthog/queries/properties_timeline/properties_timeline.py:43]: Flat is usually better than nested (see https://github.com/PostHog/posthog/issues/13315). Could we ditch the event query somehow?

File: posthog/queries/properties_timeline/properties_timeline_event_query.py [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        date_filter = f"{parsed_date_from} {parsed_date_to}"
+
+        return date_filter, query_params
+
+    def _get_lookback_date_filter(self) -> Tuple[str, Dict]:
+        query_date_range = QueryDateRange(
+            self._filter.with_data({DATE_TO: self._filter.date_from, EXPLICIT_DATE: True}),
+            self._team,
+            param_prefix="lookback",
+            is_right_open=True,  # Needs to be right-open so that this doesn't intersect with the main date range
```
Comment (macobo) [inline posthog/queries/properties_timeline/properties_timeline_event_query.py:85]: This is a no-go - we can't realisitically query ALL_OF_TIME in clickhouse with any kind of acceptable performance.

We already needed to neuter lifecycle for this reason and it will cause exact same problems here.

File: posthog/models/entity/util.py (hunk 2 above)
Comment (Twixes) [reply posthog/models/entity/util.py:9]: Do you mean a behavioral change other than support for multiple entities (logically OR'd)? This is primarily for formula mode Trends insights, where multiple entities become relevant for a single data point, which is where we need to show the properties timeline for _all_ of those entities.

File: posthog/queries/properties_timeline/properties_timeline.py (hunk 4 above)
Comment (Twixes) [reply posthog/queries/properties_timeline/properties_timeline.py:19]: `relevant_events_since_previous_point` will be shown in the timeline to inform users how many events happened between the points in time shown (since this is directly relevant to most insights).

File: posthog/queries/event_query/event_query.py (hunk 6 above)
Comment (Twixes) [reply posthog/queries/event_query/event_query.py:155]: This is just because of typing, there _is_ a `PropertiesTimelineEventQuery`, the only problem remaining at `EventQuery` level is that its `filter` needs to include `PropertiesTimelineFilter` as one of its union members, because that attribute isn't a generic (and Python generics are lackluster). So I'd prefer not to have this guard, but the guard is simpler than making `filter` generic

File: posthog/queries/query_date_range.py (hunk 7 above)
Comment (Twixes) [reply posthog/queries/query_date_range.py:250]: This is just duck typing – it was only by happenstance that previously all filter types included `IntervalMixin`, but that isn't fundamentally required for `QueryDateRange`, and indeed `PropertiesTimelineFilter` has no `interval`. Alternatively this could be an `if not isinstance(self._filter, IntervalMixin): return False` guard. What do you think?

File: posthog/queries/query_date_range.py (hunk 8 above)
Comment (Twixes) [reply posthog/queries/query_date_range.py:17]: We could, but the two new options are useful in a pretty abstract way, so I'm not convinced a new properties timeline-specific class would be clearer.

File: posthog/queries/properties_timeline/properties_timeline_event_query.py (hunk 10 above)
Comment (Twixes) [reply posthog/queries/properties_timeline/properties_timeline_event_query.py:85]: Hmm, thinking about this more it might not be that useful show the state from the before the range being queried for, so I'll actually drop this.

File: posthog/models/filters/properties_timeline_filter.py (hunk 3 above)
Comment (mariusandra) [reply posthog/models/filters/properties_timeline_filter.py:7]: Someone needs to build this backend router... 

But, keeping the backend as is (with a router in `query.ts`), could the frontend be built on top of data exploration? 🤔

File: posthog/queries/properties_timeline/properties_timeline.py (hunk 9 above)
Comment (Twixes) [reply posthog/queries/properties_timeline/properties_timeline.py:37]: Not really because window functions. At best `relevant_property_values` could be directly in the event query, but from my testing this performs about the same as the current version.

File: posthog/queries/properties_timeline/properties_timeline.py (hunk 4 above)
Comment (Twixes) [reply posthog/queries/properties_timeline/properties_timeline.py:19]: Reworked this to `relevant_event_count`, which is more relevant (events for _this_ point instead of _for previous_ point).

File: posthog/queries/properties_timeline/properties_timeline.py (hunk 9 above)
Comment (macobo) [reply posthog/queries/properties_timeline/properties_timeline.py:37]: I think we can drop this, but I'll take issue with

> performs about the same as the current version.

This is unclear without additional details. 

https://github.com/PostHog/posthog/issues/13315 Trends with nested event queries also performs identically to one where nesting is removed _until_ you have too much data and your network suddenly isn't fast enough. The indicator whether something is wrong here is network traffic for the subquery sent to the other shard.

Also note testing on team_id=2 is useless for performance optimizations - we should be looking at top 10 US clients as we're too small to be meaningful. Also also - we should be controlling for page cache :)

All this said - as long as we do proper benchmarking on this before enabling this endpoint for everyone (feature flagging the FE), I think we might be good. Happy to pair on the benchmarking of this next week.

**Review by macobo — APPROVED** [review]
:ship: Good work. Some of the inheritance code is very smelly but looks good otherwise.

Happy to unblock the backend side of this and do a benchmarking session next week.

For the frontend - could you ensure it calls `captureTimeToSeeData` when this is loaded?

File: posthog/queries/properties_timeline/properties_timeline.py (hunk 9 above)
Comment (neilkakkar) [reply posthog/queries/properties_timeline/properties_timeline.py:37]: Karl, all of these queries filter on events for a *single* person within a given time range, so even if we were just getting all rows from Clickhouse and doing aggregation in python, I think we'd be reasonably fast 😅 

(hence why I also dropped my suggestion of using arrays / going too deep in the weeds to figure out how that would work)

**Comment by Twixes** [comment]
Sure, thanks for bringing `captureTimeToSeeData()` to my attention, didn't see it before!
